cuda/preconditioner/jacobi_kernels.cu

```diff
@@ -634,14 +743,21 @@ void apply(std::shared_ptr<const CudaExecutor> exec, size_type num_blocks,
            const matrix::Dense<ValueType> *b,
            const matrix::Dense<ValueType> *beta, matrix::Dense<ValueType> *x)
 {
-    using Vec = matrix::Dense<ValueType>;
-    // TODO: do this efficiently
-    auto tmp = x->clone();
-    simple_apply(exec, num_blocks, max_block_size, storage_scheme,
-                 block_precisions, block_pointers, blocks, b,
-                 static_cast<Vec *>(tmp.get()));
-    x->scale(beta);
-    x->add_scaled(alpha, tmp.get());
+    // TODO: write a special kernel for multiple RHS
+    dense::scale(exec, beta, x);
+    for (size_type col = 0; col < b->get_size()[1]; ++col) {
+        select_advanced_apply(
+            compiled_kernels(),
+            [&](int compiled_block_size) {
+                return max_block_size <= compiled_block_size;
+            },
+            syn::value_list<int, cuda_config::min_warps_per_block>(),
+            syn::type_list<>(), num_blocks, block_precisions.get_const_data(),
+            block_pointers.get_const_data(), blocks.get_const_data(),
+            storage_scheme, alpha->get_const_values(),
```

Comment:
Since you are using `alpha[0]` inside the `apply` kernel, you also need to shift the `alpha` pointer in order to get the correct one as far as I know.
```suggestion
            storage_scheme, alpha->get_const_values() + col,
```

Reply:
Currently, I think alpha expects to always be a scalar, ever for the multiple RHS case.

See in `cuda/test/preconditioner/jacobi_kernels.cpp` the test `CudaLinearCombinationApplyToMultipleVectorsEquivalentToRef`. Cuda is a lone scalar.

Reply:
I think both cases can be the case, I only know that in the solver (like here in [`core/solver/bicgsab.cpp`](https://github.com/ginkgo-project/ginkgo/blob/develop/core/solver/bicgstab.cpp#L86)) we have as many `alpha`s as we have right hand sides.
I will look if we documented that anywhere.

Reply:
It seems that you are right, I might have confused it (the `alpha` in the solver is probably not the same as the one in `apply`).
In [the advanced `apply` in LinOp](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L527), we use `validate_application_parameters` (defined [here](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L283)), which checks if `alpha` and `beta` have `dim<2>(1,1)`.

So you are correct that it is always a `1x1` matrix.

Reply:
I did see implementations though which work with `alpha[col]`or `alpha[alpha_id]`. That is the case for CGS and some Dense kernels. There may actually be a problem there ?

Reply:
In the solvers, I was wrong, they have a separate `apply_impl` (like [here in Bicgstab](https://github.com/ginkgo-project/ginkgo/blob/develop/core/solver/bicgstab.cpp#L176)) to handle those `alpha` and `beta`. I just thought `alpha` always meant the same, which is clearly not the case (and was kind of naive from me).

Reply:
OK and in dense it's purely for scaling where alpha has the meaning of what you scale by, so that is normal also.